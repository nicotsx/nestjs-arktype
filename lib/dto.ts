import { InternalServerErrorException } from '@nestjs/common';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { ArkErrors, type Type } from 'arktype';

type Options = {
  name: string;
  input?: boolean;
};

type PrimitiveSwaggerType = 'string' | 'null' | 'number' | 'boolean' | 'integer';
type JsonSchema = { type: 'object'; properties: Record<string, JsonField>; required: string[] };
type JsonField =
  | {
      type: PrimitiveSwaggerType;
      anyOf?: never;
      const?: never;
      enum?: string[];
    }
  | {
      type: 'array';
      items: JsonField;
      anyOf?: never;
      const?: never;
    }
  | {
      type: 'object';
      properties: Record<string, JsonField>;
      required: string[];
      anyOf?: never;
      const?: never;
    }
  | {
      type?: never;
      anyOf: JsonField[];
      required?: never;
      const?: never;
    }
  | {
      const: string;
      type?: never;
      anyOf?: never;
    };

export interface ArkDto<T extends Type<unknown> = Type<unknown>> {
  new (): T['inferOut'];
  isArkDto: true;
  schema: T;
  parse(data: T['inferIn']): T['inferOut'];
  parseUnknown(data: unknown): T['inferOut'];
  input: Type<T['inferIn']>;
  output: Type<T['inferOut']>;
}

const applyApiProperties = (jsonSchema: JsonSchema, target: ArkDto) => {
  const { properties = {}, required = [] } = jsonSchema;

  for (let [key, value] of Object.entries(properties)) {
    Object.defineProperty(target.prototype, key, {
      value: undefined,
      enumerable: true,
      writable: true,
    });

    const isRequired = required.includes(key);

    if (value.anyOf?.[0]?.const) {
      value = { type: 'string', enum: value.anyOf.map((v) => v.const).filter((v) => v !== undefined) };
    }

    const Decorator = isRequired ? ApiProperty : ApiPropertyOptional;

    if (value.type === 'object' && value.properties) {
      Decorator({ selfRequired: isRequired, ...value })(target.prototype, key);
    } else {
      Decorator({ type: 'string', ...value })(target.prototype, key);
    }
  }
};

export function createArkDto<T extends Type>(schema: T, opts: Options) {
  const { name, input = false } = opts;

  class AugmentedArkDto {
    public static schema = schema;
    public static input = schema.in;
    public static output = schema.out;
    public static isArkDto = true;

    static parse(data: T['inferIn']): T['inferOut'] {
      const result = schema(data);

      if (result instanceof ArkErrors) {
        throw new InternalServerErrorException(result.summary);
      }

      return result;
    }

    static parseUnknown(data: unknown): T['inferOut'] {
      const result = schema(data);

      if (result instanceof ArkErrors) {
        throw new InternalServerErrorException(result.summary);
      }

      return result;
    }
  }

  Object.defineProperty(AugmentedArkDto, 'name', { value: name });

  const jsonSchema = input
    ? schema.in.toJsonSchema({
        fallback: { default: (ctx) => ctx.base },
      })
    : schema.out.toJsonSchema({
        fallback: { default: (ctx) => ctx.base },
      });

  applyApiProperties(jsonSchema as JsonSchema, AugmentedArkDto as unknown as ArkDto<T>);

  return AugmentedArkDto as unknown as ArkDto<T>;
}

export function isArkDto(metatype: unknown): metatype is ArkDto {
  // biome-ignore lint/suspicious/noExplicitAny: needed to check for isArkDto
  return (metatype as any)?.isArkDto;
}
