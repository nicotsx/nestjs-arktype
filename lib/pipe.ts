import { type ArgumentMetadata, BadRequestException, Injectable, type PipeTransform } from '@nestjs/common';
import { ArkErrors, type Type } from 'arktype';
import { type ArkDto, isArkDto } from './dto';

type ArkValidationPipeClass = new (dto?: ArkDto) => PipeTransform;

const validate = (value: unknown, schema: Type) => {
  const result = schema(value);

  if (result instanceof ArkErrors) {
    throw new BadRequestException(result.summary);
  }

  return result;
};

export function createArkValidationPipe(): ArkValidationPipeClass {
  @Injectable()
  class ArkValidationPipe implements PipeTransform {
    constructor(private dto?: ArkDto) {}

    public transform(value: unknown, { metatype }: ArgumentMetadata) {
      if (this.dto) {
        if (this.dto.isInput) {
          return validate(value, this.dto.input);
        }

        return validate(value, this.dto.output);
      }

      if (!metatype) {
        return value;
      }

      if (!isArkDto(metatype)) {
        return value;
      }

      if (metatype.isInput) {
        return validate(value, metatype.input);
      }

      return validate(value, metatype.output);
    }
  }

  return ArkValidationPipe;
}

export const ArkValidationPipe = createArkValidationPipe();
