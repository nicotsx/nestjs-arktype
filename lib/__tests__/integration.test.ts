import 'reflect-metadata';

import { Body, Controller, type INestApplication, Module, Post } from '@nestjs/common';
import { ApiResponse, DocumentBuilder, SwaggerModule } from '@nestjs/swagger';
import { Test } from '@nestjs/testing';
import { type } from 'arktype';
import request from 'supertest';
import { beforeEach, describe, expect, it } from 'vitest';
import { createArkDto } from '../dto';
import { ArkValidationPipe } from '../pipe';

const createUserBody = type({
  email: 'string.email',
  password: 'string',
  age: 'number = 18',
});

class CreateUserBodyDto extends createArkDto(createUserBody, { name: 'CreateUserBodyDto', input: true }) {}
class CreateUserBodyDtoOut extends createArkDto(createUserBody, { name: 'CreateUserBodyDtoOut' }) {}

@Controller()
class AppController {
  @Post('/user')
  @ApiResponse({ status: 201, type: CreateUserBodyDto })
  async createUser(@Body() body: CreateUserBodyDto): Promise<CreateUserBodyDto> {
    return CreateUserBodyDto.parse(body);
  }

  @Post('/user-out')
  @ApiResponse({ status: 201, type: CreateUserBodyDtoOut })
  async createUserOut(@Body() body: CreateUserBodyDtoOut): Promise<CreateUserBodyDtoOut> {
    return CreateUserBodyDtoOut.parse(body);
  }

  @Post('/user-in-out')
  @ApiResponse({ status: 201, type: CreateUserBodyDtoOut })
  async createUserInOut(@Body() body: CreateUserBodyDto): Promise<CreateUserBodyDtoOut> {
    return CreateUserBodyDtoOut.parse(body);
  }
}

@Module({
  controllers: [AppController],
})
class TestModule {}

describe('integration', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [TestModule],
    }).compile();

    app = moduleRef.createNestApplication();
    app.useGlobalPipes(new ArkValidationPipe());
    await app.init();
  });

  it('should correctly create the openapi schema', async () => {
    const config = new DocumentBuilder().setTitle('Runtipi API').setDescription('API specs for Runtipi').setVersion('1.0').build();

    const document = SwaggerModule.createDocument(app, config, {
      operationIdFactory: (_: string, methodKey: string) => methodKey,
    });

    expect(document).toMatchSnapshot();
  });

  it('should parse the incoming body', async () => {
    const email = 'hello@test.com';
    const password = '123456';

    return request(app.getHttpServer()).post('/user').send({ email, password }).expect(201).expect({
      email,
      password,
      age: 18,
    });
  });

  it('should throw an error if the body is invalid', async () => {
    const email = 'hello';
    const password = '123456';

    return request(app.getHttpServer()).post('/user').send({ email, password }).expect(400).expect({
      statusCode: 400,
      message: 'email must be an email address (was "hello")',
      error: 'Bad Request',
    });
  });

  it('should correctly validate against the output schema', async () => {
    const email = 'hello@hello.com';
    const password = '123456';

    return request(app.getHttpServer()).post('/user-out').send({ email, password }).expect(400).expect({
      error: 'Bad Request',
      message: 'age must be a number (was missing)',
      statusCode: 400,
    });
  });

  it('should correctly validate against the input and output schema', async () => {
    const email = 'hello@hello.com';
    const password = '123456';

    return request(app.getHttpServer()).post('/user-in-out').send({ email, password }).expect(201).expect({
      email,
      password,
      age: 18,
    });
  });
});
