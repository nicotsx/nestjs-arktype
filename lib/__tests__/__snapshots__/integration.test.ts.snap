// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration > should correctly create the openapi schema 1`] = `
{
  "components": {
    "schemas": {
      "CreateUserBodyDto": {
        "properties": {
          "age": {
            "default": 18,
            "type": "number",
          },
          "email": {
            "format": "email",
            "pattern": "^[\\w%+.-]+@[\\d.A-Za-z-]+\\.[A-Za-z]{2,}$",
            "type": "string",
          },
          "password": {
            "type": "string",
          },
        },
        "required": [
          "email",
          "password",
        ],
        "type": "object",
      },
      "CreateUserBodyDtoOut": {
        "properties": {
          "age": {
            "type": "number",
          },
          "email": {
            "format": "email",
            "pattern": "^[\\w%+.-]+@[\\d.A-Za-z-]+\\.[A-Za-z]{2,}$",
            "type": "string",
          },
          "password": {
            "type": "string",
          },
        },
        "required": [
          "age",
          "email",
          "password",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "contact": {},
    "description": "API specs for Runtipi",
    "title": "Runtipi API",
    "version": "1.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/user": {
      "post": {
        "operationId": "createUser",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/CreateUserBodyDto",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "201": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/CreateUserBodyDto",
                },
              },
            },
            "description": "",
          },
        },
        "tags": [
          "App",
        ],
      },
    },
    "/user-in-out": {
      "post": {
        "operationId": "createUserInOut",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/CreateUserBodyDto",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "201": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/CreateUserBodyDtoOut",
                },
              },
            },
            "description": "",
          },
        },
        "tags": [
          "App",
        ],
      },
    },
    "/user-out": {
      "post": {
        "operationId": "createUserOut",
        "parameters": [],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/CreateUserBodyDtoOut",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "201": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/CreateUserBodyDtoOut",
                },
              },
            },
            "description": "",
          },
        },
        "tags": [
          "App",
        ],
      },
    },
  },
  "servers": [],
  "tags": [],
}
`;
